src/shortcuts/__tests__/undo.ts

```diff
@@ -0,0 +1,294 @@
+      rankInContext: 1,
+      thoughtsRanked: [{ value: 'b', rank: 1 }]
+    },
+    // dispensible set cursor (which only updates datanonce)
+    { type: 'setCursor', thoughtsRanked: null },
+    // undo setCursor and thoughtChange in a sinle action
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+    - b
+    - c
+    - d`
+
+  expect(exported).toEqual(expectedOutput)
+})
+
+it('state remains unchanged if there are no inverse patches', () => {
+  const store = createTestStore()
+
+  store.dispatch(importText(RANKED_ROOT, `
+    - a
+     - b
+     - c
+     - d`)
+  )
+
+  const prevState = store.getState()
+  expect(prevState.inversePatches.length).toEqual(0)
+
+  store.dispatch({ type: 'undoAction' })
+
+  expect(store.getState()).toEqual(prevState)
+})
+
+it('newThought action should be merged with the succeeding patch', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b`
+    ),
+    { type: 'newThought', value: 'c' },
+    { type: 'newThought', value: 'd' },
+    {
+      type: 'existingThoughtChange',
+      context: [ROOT_TOKEN],
+      oldValue: 'd',
+      newValue: 'd1',
+      rankInContext: 3,
+      thoughtsRanked: [
+        {
+          value: 'd',
+          rank: 3,
+        }
+      ]
+    },
+    // undo thought change and preceding navigation actions
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+  - b
+  - c`
+
+  expect(exported).toEqual(expectedOutput)
+
+})
+
+it('undo contiguous changes', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - A
+      - B`
+    ),
+    {
+      type: 'existingThoughtChange',
+      newValue: 'Atlantic',
+      oldValue: 'A',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'A', rank: 0 }]
+    },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'Atlantic ',
+      oldValue: 'Atlantic',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'Atlantic', rank: 0 }]
+    },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'Atlantic City',
+      oldValue: 'Atlantic ',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'Atlantic ', rank: 0 }]
+    },
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - A
+  - B`
+
+  expect(exported).toEqual(expectedOutput)
+
+})
+
+it('state.alert is omitted from the undo patch', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - A
+      - B`
+    ),
+    { type: 'alert', value: 'test' }
+  ])
+
+  expect(store.getState().inversePatches.length).toEqual(0)
+
+})
```

Comment:
I'm not sure if that's the correct approach for writing this unit test. Because we're assuming that `archiveThought` would always include `alert`. If the alert is excluded from the `archiveThought`, the test would still pass even if the `undo reducer` logic is incorrect.

Reply:
@anmolarora1 The test should assert that `state.alert` is not restored after an undo when it has been dismissed. This would fail if it weren't for `stateSectionsToOmit`. Effectively we need a separate test for `stateSectionsToOmit` since it is different than just skipping a non-undoable action.

Reply:
That's correct. 
And I see two approaches for handling this - 1) We write our test using an existing reducer which uses `state.alert`, e.g. `archiveThought`. But as I mentioned [here](https://github.com/cybersemics/em/pull/811#discussion_r498657964), it relies on the assumption that the reducer wouldn't change.
2) We use a `test/mock reducer` and a `mock state`. Afaik, it's not possible with the existing `testStore`, and so we'll need to update it to allow passing a mock reducer as an argument instead of using the [appReducer](https://github.com/cybersemics/em/blob/5be62eb42d613aade4fecdecc70974beb16414d8/src/test-helpers/createTestStore.js#L13)
Besides, we'd need to inspect the patches that are generated to verify that they don't contain `alert` related information. 
What are your thoughts?

Reply:
The first approach seems much more straightforward. I don't mind having a dependency on `archiveThought` this time.